### PR TITLE
Feat rccp interface

### PR DESCRIPTION
**Pull request by msupernaw**
Thanks for opening a pull request (PR) to FIMS! Please describe the changes under the headings below. Please also go through the checklist below to ensure the appropriate steps have been taken during development. For smaller changes, feel free to skip steps that aren't relevant, but at least describe the feature under the "what is the feature".

# What is the feature?
A brief title of this change.

# How have you implemented the solution?
Describe the changes.

# Does it impact any other area of the project?
Please describe which features are impacted by this change.

# Does this change impact the model input or output?*
Please describe any way this change impacts the model input or output.

# How to test this change
Please include a test file and/or Github workflow.

# Developer pre-PR Checklist

Please do these steps locally for big changes. Note GitHub Actions does all of these checks automatically when pushing to the repository. Please see [code development section in the contributor guide](https://noaa-fims.github.io/collaborative_workflow/contributor-guidelines.html#code-development) for details.

- [ ] Run [cmake build and ctest locally](https://noaa-fims.github.io/collaborative_workflow/testing.html#c-unit-testing-and-benchmarking) and make sure the C++ tests pass
- [ ] Run `devtools::document()` locally and push changes to the remote feature branch
- [ ] Run `styler::style_pkg()` locally and push changes to remote feature branch. If there are many changes, please do this in a separate commit.
- [ ] Run `devtools::check()` locally and make sure the package can be compiled and R tests pass. If there are failing tests, run `devtools::test(filter = "file_name")` (where "test-file_name.R" is the testthat file containing failing tests) and edit code/tests to troubleshoot tests.
- [ ] Before opening the PR, make sure all the github actions are passing on the remote feature branch.
- [ ] Make sure this PR has an informative title.

### CODE REVIEW

File: inst/include/common/information.hpp [hunk 1]
```diff
@@ -24,16 +25,45 @@ namespace fims {
...
+  std::vector<T*>
       random_effects_parameters;  // list of all random effects parameters
-  std::vector<variable_t*>
+  std::vector<T*>
       fixed_effects_parameters;  // list of all fixed effects parameters
+  
+  std::map<uint32_t, std::shared_ptr<fims::RecruitmentBase<T> > > recruitment_models;
+  std::map<uint32_t, std::shared_ptr<fims::SelectivityBase<T> > > selectivity_models;
+  
+static std::shared_ptr<Information<T> > get_instance(){
```
Comment (ChristineStawitz-NOAA) [inline inst/include/common/information.hpp:41]: per Google style guides, function names should be CamelCase e.g. GetInstance()

File: inst/include/common/information.hpp [hunk 2]
```diff
@@ -24,16 +25,45 @@ namespace fims {
...
+   if(Information<T>::fims_information == nullptr){
+       Information<T>::fims_information =  std::make_shared<fims::Information<T> >();
+   }  
+   return Information<T>::fims_information;
 
-  bool CreateModel() {}
+}
+
+void register_parameter( T& p){
```
Comment (ChristineStawitz-NOAA) [inline inst/include/common/information.hpp:49]: Should be RegisterParameter

File: inst/include/common/information.hpp [hunk 3]
```diff
@@ -24,16 +25,45 @@ namespace fims {
...
+    this->parameters.push_back(&p);
+}
+
+void register_random_effect( T& re){
```
Comment (ChristineStawitz-NOAA) [inline inst/include/common/information.hpp:53]: Should be RegisterRandomEffect()

File: inst/include/common/information.hpp [hunk 4]
```diff
@@ -24,16 +25,45 @@ namespace fims {
...
+    this->random_effects_parameters.push_back(&re);
+}
+
+ bool creat_model() {
```
Comment (ChristineStawitz-NOAA) [inline inst/include/common/information.hpp:57]: Should be CreateModel()

File: inst/include/interface/rcpp/rcpp_interface.hpp [hunk 5]
```diff
@@ -0,0 +1,512 @@
+#ifndef FIMS_INTERFACE_RCPP_INTERFACE_HPP
+#define FIMS_INTERFACE_RCPP_INTERFACE_HPP
+
+/*
+ * File:   rcpp_interface.hpp
```
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/rcpp/rcpp_interface.hpp:5]: use standard FIMS disclaimer

File: inst/include/interface/rcpp/rcpp_interface.hpp [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+
+
+#define RCPP_NO_SUGAR
+#include <Rcpp.h>
+
+/**
+ * @brief RcppInterface class that defines 
+ * the interface between R and C++ for parameter types.
+ */
+class parameter {
```
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/rcpp/rcpp_interface.hpp:51]: should be Parameter

File: inst/include/interface/rcpp/rcpp_interface.hpp [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+
+    parameter() {
+        this->value = 0;
+    }
+};
+
+/**
+ *@brief Base class for all interface objects
+ */
+class fims_rcpp_interface_base {
```
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/rcpp/rcpp_interface.hpp:78]: Should be FIMSRcppInterfaceBase

File: inst/include/interface/rcpp/rcpp_interface.hpp [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+        std::cout << "fims_rcpp_interface_base::add_to_fims_tmb(): Not yet implemented.\n";
+        return false;
+    }
+};
+std::vector<fims_rcpp_interface_base*> fims_rcpp_interface_base::fims_interface_objects;
+
+/****************************************************************
+ * Recruitment Rcpp interface                                   *
+ ***************************************************************/
+class recruitment_interface_base : public fims_rcpp_interface_base {
```
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/rcpp/rcpp_interface.hpp:92]: Should be RecruitmentInterfaceBase

File: inst/include/interface/rcpp/rcpp_interface.hpp [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+
+};
+
+uint32_t recruitment_interface_base::id_g = 1;
+std::map<uint32_t, recruitment_interface_base* > recruitment_interface_base::recruitment_objects;
+
+/**
+ * @brief Interface class for Beverton-Holt recruitment. 
+ */
+class beverton_holt : public recruitment_interface_base {
```
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/rcpp/rcpp_interface.hpp:112]: The FIMS class is SRBevertonHolt so in additional to camel casing I would call this something like SRBevertonHoltInterface

File: inst/include/interface/rcpp/rcpp_interface.hpp [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+    beverton_holt() {
+        this->id = recruitment_interface_base::id_g++;
+        recruitment_interface_base::recruitment_objects[this->id] = this;
+        fims_rcpp_interface_base::fims_interface_objects.push_back(this);
+    }
+
+    virtual ~beverton_holt() {
+    }
+
+    virtual bool add_to_fims_tmb() {
```
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/rcpp/rcpp_interface.hpp:128]: Should be AddToFIMSTMB

File: inst/include/interface/rcpp/rcpp_interface.hpp [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+        d3->recruitment_models[b3->id] = b3;
+
+        return true;
+    }
+};
+
+/****************************************************************
+ * Selectivity Rcpp interface                                   *
+ ***************************************************************/
+class selectivity_interface_base : public fims_rcpp_interface_base {
```
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/rcpp/rcpp_interface.hpp:287]: Should be SelectivityInterfaceBase

Also I have a question here - do we need one interface for each type of function (e.g. Selectivity) or does there need to be one for each module, e.g. LogisticSelectivityInterface

File: inst/include/interface/rcpp/rcpp_interface.hpp [hunk 12]
```diff
@@ -0,0 +1,512 @@
...
+
+};
+
+uint32_t growth_interface_base::id_g = 1;
+std::map<uint32_t, growth_interface_base* > growth_interface_base::growth_objects;
+
+/****************************************************************
+ * Maturity Rcpp interface                                      *
+ ***************************************************************/
+class maturity_interface_base : public fims_rcpp_interface_base {
```
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/rcpp/rcpp_interface.hpp:332]: Should be MaturityInterfaceBase

File: inst/include/interface/rcpp/rcpp_interface.hpp [hunk 13]
```diff
@@ -0,0 +1,512 @@
...
+
+};
+
+uint32_t natural_mortality_base::id_g = 1;
+std::map<uint32_t, natural_mortality_base* > natural_mortality_base::natural_mortality_objects;
+
+/****************************************************************
+ * Population Rcpp interface                                    *
+ ***************************************************************/
+class population : public fims_rcpp_interface_base {
```
Comment (ChristineStawitz-NOAA) [inline inst/include/interface/rcpp/rcpp_interface.hpp:368]: Population should be capitalized

File: tests/FIMS_Rcpp_test.R [hunk 14]
```diff
@@ -0,0 +1,18 @@
+
```
Comment (ChristineStawitz-NOAA) [inline tests/FIMS_Rcpp_test.R:1]: please put in a testthat case e.g. as in testplaceholder

**Review by ChristineStawitz-NOAA — CHANGES_REQUESTED** [review]
- [ ] Per Google C++ style, class definitions and functions should be CamelCase
- [ ] the R testing function should follow testthat conventions
- [x] it's unclear whether each of the interfaces should be at the functor level e.g. beverton holt or the folder level e.g. recruitment, please be consistent.

File: inst/include/common/information.hpp (hunk 1 above)
Comment (msupernaw) [reply inst/include/common/information.hpp:43]: Class method and function names throughout the FIMS codebase use lowercase/underscore conventions. Class names are using CamelCase, should we refactor the whole codebase?

**Comment by msupernaw** [comment]
"it's unclear whether each of the interfaces should be at the functor level e.g. beverton holt or the folder level e.g. recruitment, please be consistent." I don't understand what you are saying here?

**Comment by ChristineStawitz-NOAA** [comment]
> "it's unclear whether each of the interfaces should be at the functor level e.g. beverton holt or the folder level e.g. recruitment, please be consistent." I don't understand what you are saying here?

Do we need a new Interface Base for each derived class, for example, we would need like a BevertonHoltInterfaceBase AND a RickerInterfaceBase, or do we just need an InterfaceBase for the parent class, for example RecruitmentInterfaceBase?

**Comment by msupernaw** [comment]
Just a module specific base that extends the FIMSRCppInterfaceBase RecruitmentInterfaceBase.

**Review by ChristineStawitz-NOAA — APPROVED** [review]

**Pull request merged**